dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEnpointChannels2.java

```diff
@@ -0,0 +1,194 @@
+        this.delegate = Preconditions.checkNotNull(endpointChannelFactory, "endpointChannelFactory");
+    }
+
+    @Override
+    public Channel get() {
+        return new StickyChannel2(delegate.get());
+    }
+
+    @Override
+    public String toString() {
+        return "StickyEndpointChannels2{" + delegate + "}";
+    }
+
+    private static final class StickyChannel2 implements EndpointChannelFactory, Channel {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router = new StickyRouter();
+
+        private StickyChannel2(EndpointChannelFactory channelFactory) {
+            this.channelFactory = channelFactory;
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            return new StickyEndpointChannel(router, channelFactory.endpoint(endpoint));
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+            return endpoint(endpoint).execute(request);
+        }
+
+        @Override
+        public String toString() {
+            return "Sticky{" + channelFactory + '}';
+        }
+    }
+
+    public static StickyEnpointChannels2 create(Supplier<EndpointChannelFactory> endpointChannelFactory) {
+        return new StickyEnpointChannels2(endpointChannelFactory);
+    }
+
+    private static final class StickyRouter {
+
+        private final InFlightCallSuccessTransformer successTransformer = new InFlightCallSuccessTransformer();
+        private final InFlightCallFailureTransformer failureTransformer = new InFlightCallFailureTransformer();
+
+        @Nullable
+        private volatile StickyTarget stickyTarget;
+
+        @Nullable
+        @GuardedBy("this")
+        private volatile ListenableFuture<Response> callInFlight;
+
+        public ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel) {
+            if (stickyTarget != null) {
+                return executeWithStickyTarget(stickyTarget, request, endpointChannel);
+            }
+
+            synchronized (this) {
+                if (stickyTarget != null) {
+                    return executeWithStickyTarget(stickyTarget, request, endpointChannel);
+                }
+
+                ListenableFuture<Response> callInFlightSnapshot = callInFlight;
+                if (callInFlightSnapshot == null) {
+                    // Cannot use DialogueFutures#addDirectCallback because we want ordering of listeners:
+                    // we want the success/failure callbacks to run BEFORE the queued requests inspect the result of
+                    // the first call.
+                    ListenableFuture<Response> result = DialogueFutures.transform(
```

Comment:
@carterkozak I originally had this as a callback, but I couldn't convince myself that's safe: ListenableFutures do not guarantee order of callbacks. I guess it's safe because queued callbacks that run before the in-flight callback will just queue themselves again? Would that just OOM, because those callbacks would execute immediately? I think I'm not familiar enough with ListenableFutures impl, so this feels safer.

Reply:
The ListenableFuture reference implementation notes that they cannot change the order because it would break a bunch of code internally and externally, but it's still better to transform when we require ordering because it's easier to reason about.